### ISSUE TITLE
raft: server_impl: wait_for_state_change can be behind is_leader

### ISSUE THREAD

**Issue by patjed41** [scylladb/scylladb#15784]
Consider such code:
```cpp
if (server.is_leader()) {
    co_await server.wait_for_state_change();
}
```
Currently, it is possible that if `server.is_leader()` returns true, we finish `co_await server.wait_for_state_change();` without an actual state change. It can happen because `wait_for_state_change` can be "behind" `is_leader` and detect a state change to leader that happened before calling `is_leader`.

The explanation for this behavior is simple. `server_impl::is_leader` uses `fsm` directly to check if the server is a leader. On the other hand, `server_impl::wait_for_state_change` uses a promise whose value is set by `server_impl::io_fiber` when it receives `fsm_output` with a state change from `fsm`. It is possible that `wait_for_state_change` is called after the server becomes a leader according to `fsm`, but before `io_fiber` polls a corresponding output.

For now, the only consequence of this behavior I noticed is in the logs printed by `raft_group0::leadership_monitor_fiber`:
```
INFO  2023-10-20 16:02:37,171 [shard 0:stre] raft_group0 - gaining leadership
INFO  2023-10-20 16:02:37,172 [shard 0:stre] raft_group0 - losing leadership
INFO  2023-10-20 16:02:37,172 [shard 0:stre] raft_group0 - gaining leadership
```
It happens because `wait_for_state_change` that should catch leadership loss actually catches leadership gain from earlier. These are only logs, but we use the same logic -- `server.is_leader()` returns true -> do `co_await server.wait_for_state_change();` -> assume we are not a leader anymore -- in other places like `topology_coordinator::step_down_as_nonvoter`. We should synchronize `is_leader` and `wait_for_state_change` to avoid problems in the future.

Apart from `_state_change_promise`, there are more similar promises in `server_impl`. For now, I'm not sure if all of them are used correctly or there is a similar problem.

**Comment by kbr-scylla** [comment]
cc @gleb-cloudius

**Comment by gleb-cloudius** [comment]
I think this is benign. It is futile to try to be precise here since the state changes asynchronously.  The code needs to recheck that state after `wait_state_change` anyway. What is important that no state changes are missed.

**Comment by kbr-scylla** [comment]
> What is important that no state changes are missed.

IIUC we are missing state changes, because `wait_for_state_change` returns after detecting a "past" state change. It should return on the first future state change.

so in the code:
```
if (X) wait_for_state_change();
```

`wait_for_state_change` should not return for the state change that caused `X` to be true, it should return for a later state change.

**Comment by gleb-cloudius** [comment]
> > What is important that no state changes are missed.
> 
> IIUC we are missing state changes, because `wait_for_state_change` returns after detecting a "past" state change. It should return on the first future state change.
> 
> so in the code:
> 
> ```
> if (X) wait_for_state_change();
> ```
> 
> `wait_for_state_change` should not return for the state change that caused `X` to be true, it should return for a later state change.

I do not understand. Can you give more specific example. It is fine to coalesce state changes of course. I think this is what you describe in your example.

**Comment by gleb-cloudius** [comment]
IOW this is level triggered event.

**Comment by kostja** [comment]
I agree with Gleb's assessment, this is exactly the discussion we had over one of his pull requests. So what is the real issue here?

**Comment by kbr-scylla** [comment]
`wait_for_state_change` may return without the state changing. I think it's confusing.

**Comment by gleb-cloudius** [comment]
> `wait_for_state_change` may return without the state changing. I think it's confusing.

This is how level triggered evens behave. There was state change, but you called `wait_for_state_change` after it happened. It may also happen that a multiple events were coalesced and the `wait_for_state_change`  will trigger though the state will be the same before and after. And this is also fine.

**Comment by kbr-scylla** [comment]
Both the name of the function and the comment imply that after `wait_for_state_change()` returns, state will change.
```cpp
    // Returned future is resolved when state changes
    // State changes can be coalesced, so it is not guarantied that the caller will
    // get notification about each one of them. The state can even be the same after
    // the call as before, but term should be different.
    virtual future<> wait_for_state_change(seastar::abort_source* as = nullptr) = 0;
```
If it behaves differently, something has to change:
- name and comment
- or behavior

Because right now it is unreliable.

If we can easily fix the behavior (which IIUC we can, we already discussed this with @patjed41 and the solution is simple), I think we should.

**Comment by gleb-cloudius** [comment]
> Both the name of the function and the comment imply that after `wait_for_state_change()` returns, state will change.
> 
> ```c++
>     // Returned future is resolved when state changes
>     // State changes can be coalesced, so it is not guarantied that the caller will
>     // get notification about each one of them. The state can even be the same after
>     // the call as before, but term should be different.
>     virtual future<> wait_for_state_change(seastar::abort_source* as = nullptr) = 0;
> ```
> 
> If it behaves differently, something has to change:
> 
>     * name and comment
> 
>     * or behavior
> 
> 
> Because right now it is unreliable.

In what way it is not? The comment states that the state may stay the same after the function returns which is exactly what happen in the example. 

> 
> If we can easily fix the behavior (which IIUC we can, we already discussed this with @patjed41 and the solution is simple), I think we should.

I do not think we need to complicate simple and well understood level triggered even behaviour. If you would write a driver for a HW you would see it often since interrupts and HW state changes are coming from two different channels.

**Comment by kbr-scylla** [comment]
> In what way it is not? The comment states that the state may stay the same after the function returns which is exactly what happen in the example.

It also says "Returned future is resolved **when state changes**." Which is false.

> I do not think we need to complicate simple and well understood level triggered even behaviour. 

I don't propose to complicate, I propose to simplify. The function specification is simple, let's make it behave according to the spec.

BTW. I never heard of "level triggered event behavior." I googled it now, and I'm having trouble finding a simple definition in programming terms that will be directly applicable to the situation we have here.

Even if it was "well understood", the function's comment doesn't make any reference to level triggered event behavior. It specifies something else from what the function does.

**Comment by gleb-cloudius** [comment]
> > In what way it is not? The comment states that the state may stay the same after the function returns which is exactly what happen in the example.
> 
> It also says "Returned future is resolved **when state changes**." Which is false.

It is not, because it does not say that the change couldn't already happen.

> 
> > I do not think we need to complicate simple and well understood level triggered even behaviour.
> 
> I don't propose to complicate, I propose to simplify. The function specification is simple, let's make it behave according to the spec.

The spec is fine and matches the function. The code in question assumes that each function call corresponds to a state change which is wrong assumption.

> 
> BTW. I never heard of "level triggered event behavior." I googled it now, and I'm having trouble finding a simple definition in programming terms that will be directly applicable to the situation we have here.
>

Search for level triggered interrupts. 
 
> Even if it was "well understood", the function's comment doesn't make any reference to level triggered event behavior. It specifies something else from what the function does.

The function clearly states that the state may not change after the function call. If you think any reason why it may happen needs to be written specifically you are free to add it, but this will be a waist of bits IMO. The function documentation is specific enough so that it can be used correctly.

**Comment by gleb-cloudius** [comment]
If you do not like level triggered event analogy then think about condition variable spurious wakeups. It is normal to have then in CV implementation and there is no need to explain why it may happen, only that it can.

**Comment by kostja** [comment]
Gleb, on one part I agree with @kbr-scylla - this place level-triggered two people already. There is something wrong with it.
